Library/Homebrew/dependency.rb

```diff
@@ -145,3 +145,31 @@ def installed?
     false
   end
 end
+
+class PrioritizedTapDependency < Dependency
+  attr_reader :tap
+
+  def initialize(name, tags=[], env_proc=DEFAULT_ENV_PROC, option_name=name)
+    @tap, _, option_name = option_name.rpartition "/"
```

Comment:
I see the problem here. To simplify it, I think we can do something like this:
* Let's use `owner_tap` instead of `tap` as the attribute name, so it won't be confused with `tap` attribute from `TapDependency`.
* Let's make `owner_tap` as a common attributes for all `Dependency` class. So you only need to update the `Dependency.merge_repeats` method instead of creating a new one.
* I don't think we need another class of `PrioritizedTapDependency`. You can just make necessary change in `Dependency` class itself, such as:
```ruby
class Dependency

  def initialize(name, tags = [], env_proc = DEFAULT_ENV_PROC, option_name = name, owner_tap = nil) 
    ...
  end

  def to_formula
    if owner_tap.nil? || name ~= HOMEBREW_TAP_FORMULA_REGEX
      formula = Formulary.factory(name)
    else
      begin
        formula = Formulary.factory("#{owner_tap}/#{name}")
      rescue FormulaUnavailableError
        formula = Formulary.factory(name)
        unless formula.core_formula?
          opoo "<warning message>"
        end
      end
    end
    formula.build = BuildOptions.new(options, formula.options)
    formula
  end
end
```

In this approach, if owner_tap is nil, it means we don't care the where formula is initialize from. What do you think?